docs(ast): auto-generate doc comments for AstBuilder methods

DonIsaac opened the pull request:
# What This PR Does

Modifies `ast_codegen` to auto-generate rustdoc comments for generated `AstBuilder` methods. As we add more doc comments to AST node fields, the generated documentation will get better.

![image](https://github.com/user-attachments/assets/d27f0d53-38dd-4ba8-93e0-ffaf5c8e6809)

DonIsaac commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4471?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4471** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4471?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4413** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4413?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented on crates/oxc_ast/src/ast/ts.rs:
```diff
@@ -96,16 +123,43 @@ pub enum TSEnumMemberName<'a> {
+/// const x: 'foo' = 'foo';
+/// //       ^^^^^
+///
+/// type NonZero<N> = N extends 0 ? never : N;
+/// //                          ^
```
wrong pointing

DonIsaac replied:
![image](https://github.com/user-attachments/assets/4814b859-8ef7-49b9-baf8-5881fd2de349)

Dunqing commented:
Looks good to me. Add these docs very useful for contributors who are new to `oxc`!  Also, I have noticed #2803 is doing the same thing. @Boshen cc

Boshen reviewed (commented):
LGTM

rzvxa commented on tasks/ast_codegen/src/generators/ast_builder.rs:
```diff
@@ -309,6 +347,139 @@ impl ToTokens for Param {
 }
 
+#[derive(Debug)]
+struct DocComment<'p> {
```
It'll be nice if we move this type to somewhere else so we can use it across the board. I haven't spent time documenting the generated code but this is a nice way to do it.

DonIsaac replied:
Where would you like it moved? And could this be done in a following PR?

rzvxa commented:
This is an amazing addition, LGTM. We can move the `DocComment` to somewhere else in an upcoming PR.

rzvxa commented on crates/oxc_ast/src/generated/ast_builder.rs:
```diff
@@ -3489,16 +5470,36 @@ impl<'a> AstBuilder<'a> {
+    /// ## Parameters
+    /// - kind
+    /// - type_annotation
+    /// - optional
```
It would be super cool if we add an example here to let the user know what is the ideal usage(especially when some optional arguments are None).

e.g.
```rust
let id = builder.binding_pattern(
    binding_kind,
    Option::<TSTypeAnnotation>::None,
    false
);
```

The compiler tells the user to do this which is less than ideal.

```rust
let id = builder.binding_pattern::<Option<TSTypeAnnotation>>(binding_kind, None, false);
```

rzvxa replied:
I just want to add a lot of issues with lots of booleans and optional types would go away as we move toward traced constructors and template macros.
Let me know if you have any suggestions to help with the interface.

DonIsaac replied:
I agree with the examples. I think this could be tackled in a following PR, so I'm resolving this comment.

overlookmotel commented:
A couple of LGTM comments, but no-one has merged? Is there any reason not to at this point? @rzvxa I think you're the one who's most on top of this?

DonIsaac committed:
docs(ast): auto-generate doc comments for AstBuilder methods (#4471)

# What This PR Does

Modifies `ast_codegen` to auto-generate rustdoc comments for generated `AstBuilder` methods. As we add more doc comments to AST node fields, the generated documentation will get better.

![image](https://github.com/user-attachments/assets/d27f0d53-38dd-4ba8-93e0-ffaf5c8e6809)
